Worker Indexing Changes

anandagopal6 opened the pull request:
<!-- Please provide all the information below.  -->

### Issue describing the changes in this PR

https://github.com/Azure/azure-functions-host/issues/7496 

### Wiki

https://github.com/Azure/azure-functions-host/wiki/Worker-Indexing-Changes

Design doc - [Here](https://microsoft.sharepoint.com/:w:/r/teams/Antares/_layouts/15/Doc.aspx?sourcedoc=%7BE0283F6F-8072-4173-9D50-76966BD6D8F2%7D&file=Project%20Stein%20(Worker%20Indexing)%20Reference%20Documentation.docx&wdLOR=c0AEA68A1-EC24-47E6-80AA-D5F67612F840&action=default&mobileredirect=true&cid=605d80a5-d22d-40c3-abfd-cd3301d2ba60)

### Pull request checklist

* [ ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)

<!-- Optional: delete if not applicable  -->
### Additional information

Additional PR information

anandagopal6 committed:
Merge dev into branch containing all worker indexing code changes/additions.

anandagopal6 committed:
Fix test failures by manually adding worker indexing environment variables/worker config variables to test cases.

Add null checks to _workerRuntime and workerConfig code sections.

anandagopal6 committed:
Fix test failures related to FunctionMetadataProviderFactory initialization.

anandagopal6 committed:
Remove extra line of code in GrpcWorkerChannel and try something for integration tests

anandagopal6 committed:
Remove unnecessary workerIndexing variable from FunctionMetadataManager

anandagopal6 committed:
Remove experimental line

anandagopal6 committed:
Fix build issue by removing unnecessary variable

anandagopal6 committed:
Add experimental line to see if this fixes integration/non-integration tests

anandagopal6 committed:
Removed _channels from RpcFunctionInvocationDispatcher, fixed boolean logic in FunctionMetadataProviderFactory

anandagopal6 committed:
Renaming GRPC messages, fix validation and FinishInitialization

anandagopal6 committed:
Renamed GrpcWorkerChannel worker indexing functions to match GRPC message names

Fixed invocation troubles for worker indexed functions --> was just missing a few lines from ScriptHost

anandagopal6 committed:
Removed WorkerCapabilities service

anandagopal6 committed:
Adding utility function for checking worker indexing and also added lots of tests

Replaced dispatcher.InitializeAsync(null) with new List<FunctionMetadata> as the parameter instead of null

Renamed FunctionMetadataProviderTests as HostFunctionMetadataProviderTests

soninaren committed:
renaming a some methods and variables, changing some log messages

soninaren committed:
Removed the dispatcher from GetFunctionMetadata() method, removed the provider factory, updated the tests

anandagopal6 committed:
Get the worker config based on runtime language

soninaren committed:
Updated Utility function, skip extension filtering if worker indexing is enabled

soninaren committed:
fixing tests

anandagopal6 committed:
Fixed some Utility.CanWorkerIndex tests by adding env variable to testEnv

Added new RawFunctionMetadata type

Moved Binding parsing logic into WorkerFunctionMetadataProvider

anandagopal6 committed:
Added retry options and configuration source to RpcFunctionMetadata and ValidateMetadata() validation logic

Added simple HttpFunctionInvocationDispatcher implementation for worker indexing functions

Cleaned ScriptHost worker indexing logic and WorkerFunctionMetadataProvider logic

anandagopal6 committed:
Cleaning up some code after reviewing the diff comparison again

anandagopal6 committed:
Add clarifying comments to new RpcFunctionMetadata attributes

anandagopal6 committed:
Rename variable in protobuf

anandagopal6 committed:
Add bundles test for bypassing filtering when worker is indexing

Add detailed binding validation logic to WorkerFunctionMetadataProvider

Add detailed comments to HTTP dispatcher

anandagopal6 committed:
Add binding validation tests

Throw NotSupportedException in HttpFunctionInvocationDispatcher for worker indexing-related functions

satvu commented on src/WebJobs.Script/Host/HostFunctionMetadataProvider.cs:
```diff
@@ -12,6 +12,7 @@
 using Microsoft.Azure.WebJobs.Script.Description;
 using Microsoft.Azure.WebJobs.Script.Diagnostics;
 using Microsoft.Azure.WebJobs.Script.Diagnostics.Extensions;
+using Microsoft.Azure.WebJobs.Script.Workers;
```
nit: not used in this file

satvu commented on src/WebJobs.Script/Host/WorkerFunctionMetadataProvider.cs:
```diff
@@ -0,0 +1,190 @@
+
+                    function.Language = SystemEnvironment.Instance.GetEnvironmentVariable(RpcWorkerConstants.FunctionWorkerRuntimeSettingName);
+
+                    // skip function directory validation because this involves reading function.json
```
nit: do we need to keep these comments?

satvu commented on src/WebJobs.Script/Workers/Http/HttpFunctionInvocationDispatcher.cs:
```diff
@@ -12,6 +12,7 @@
 using Microsoft.Azure.WebJobs.Script.Description;
 using Microsoft.Azure.WebJobs.Script.Diagnostics;
 using Microsoft.Azure.WebJobs.Script.Eventing;
+using Microsoft.Azure.WebJobs.Script.Workers.Rpc;
```
nit: remove unnecessary import

satvu commented on src/WebJobs.Script/Workers/Http/HttpFunctionInvocationDispatcher.cs:
```diff
@@ -201,5 +202,21 @@ public async Task<bool> RestartWorkerWithInvocationIdAsync(string invocationId)
             return true;
         }
+
+        public Task<IEnumerable<RawFunctionMetadata>> GetWorkerMetadata()
```
context question: what is the purpose of this method and the next (finish initialization)?

~~is it in case someone tries to use this invocation dispatcher instead of RpcFunctionInvocationDispatcher? Why choose to make the method available to call but immediately throw an exception?~~  I think the error message or comments could be made clearer b/c I don't immediately understand what is happening here.

edit: scrolled down to the next file and saw that the interface was changed.

surgupta-msft replied:
I could not follow this comment. Please share more details on it.

satvu commented on src/WebJobs.Script/Workers/IFunctionInvocationDispatcher.cs:
```diff
@@ -2,10 +2,12 @@
 // Licensed under the MIT License. See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Concurrent;
```
nit: this and line 10 are unnecessary imports

satvu commented on src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:
```diff
@@ -10,6 +10,7 @@
 using System.Threading.Tasks;
 using System.Threading.Tasks.Dataflow;
 using Microsoft.AspNetCore.Hosting;
+using Microsoft.Azure.WebJobs.Script.Config;
```
nit: unnecessary import (if these import comments are wrong, i may need to check on my VS configuration)

alrod reviewed (changes requested):
@anandagopal6, can you please add some information about the change in the PR description? why do we need it? is there a github item for the change?

satvu commented:
> @anandagopal6, can you please add some information about the change in the PR description? why do we need it? is there a github item for the change?

@alrod I don't know if she still has access as a summer intern, but these changes are related to Project Stein. It adds worker-driven function indexing. [Wiki page here](https://github.com/Azure/azure-functions-host/wiki/Worker-Indexing-Changes). 

I believe @surgupta-msft is owning this PR now?

satvu commented on test/WebJobs.Script.Tests/ScriptHostTests.cs:
```diff
@@ -17,6 +17,7 @@
 using Microsoft.Azure.WebJobs.Script.Description;
 using Microsoft.Azure.WebJobs.Script.Diagnostics;
 using Microsoft.Azure.WebJobs.Script.Eventing;
+using Microsoft.Azure.WebJobs.Script.Tests.Workers.Rpc;
```
nit: remove unnecessary import. i don't think the import below is needed either.

satvu commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -428,64 +428,72 @@ internal async Task SendInvocationRequest(ScriptInvocationContext context)
-            var functions = new List<FunctionMetadata>();
-
-            foreach (var metadata in workerFunctionMetadataResponse.Results)
+            foreach (var metadataResponse in functionMetadataResponses.Results)
```
Note that this will need to change to `functionMetadataResponses.FunctionLoadRequestsResults`. 

[PR link](https://github.com/Azure/azure-functions-language-worker-protobuf/pull/59) for the grpc message change from `Results` to `FunctionLoadRequestsResults`.

surgupta-msft replied:
Done. Thanks for sharing info here.

satvu commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -428,64 +428,72 @@ internal async Task SendInvocationRequest(ScriptInvocationContext context)
         {
             _workerChannelLogger.LogDebug("Received the worker function metadata response from worker {worker_id}", _workerId);
 
-            var responseStatus = workerFunctionMetadataResponse.OverallStatus;
```
Since OverallStatus is never used/called in the host, do we still need to have it defined in the language worker protobuf? I notice that it was replaced by checking the status of each metadata object a few lines down. Can I get some context on this decision and whether or not it could be useful in the future to keep this property?

soninaren replied:
It was added as extensibility point. It may not be read by host today but we can certainly do so in future. One example of overall status failure would `unable to access storage` or `path not found`. This is where the entire request failed because of a fundamental error.

surgupta-msft commented:
> @anandagopal6, can you please add some information about the change in the PR description? why do we need it? is there a github item for the change?

Added links pointing to github issue and design doc in the description.

surgupta-msft committed:
Merge with latest Dev branch

surgupta-msft committed:
Sending runtime when invoking GetTestFunctionDispatcher

surgupta-msft committed:
Making changes related to new proto file changes

surgupta-msft committed:
Removing unnecessary imports

surgupta-msft committed:
Removing blank line

surgupta-msft committed:
Making forceRefresh as true in GetFunctionsMetadata

Reinitializing functions array

Setting env variable in test

Removing env variable from tests

debugging integration tests

Integration tests

Debugging integration tests

Setting env variable for dotnet

Env as node for test OnTimeoutException_OOP_HasExpectedLogs

minor debug point

minor

Fixing test Validate_HostLogs

Added extra logs for debugging tests

Commenting extra log lines for test to pass

surgupta-msft committed:
Making IFunctionMetadataProvider.GetFunctionMetadata() an async method (#7796)

* Making IFunctionMetadataProvider.GetFunctionMetadata() an async method

* Adding await Task.FromResult

* Correcting the await operation

* Removing duplicate GetFunctionMetadataAsync method

* Minor fix

surgupta-msft committed:
Worker Indexing - Moving common validate logic to utility (#7807)

* Moving ValidateBinding to Utility

* Moving ValidateFunction to utility and modifying tests to remove dupliation

* Fixing PR comments

* Removing proxy from message

surgupta-msft committed:
Merge branch 'dev' into t-anjanan/workingSteinChanges-fork

surgupta-msft committed:
Merge branch 't-anjanan/workingSteinChanges-fork' of https://github.com/Azure/azure-functions-host into t-anjanan/workingSteinChanges-fork

surgupta-msft committed:
Removing extra logs used for debugging

surgupta-msft committed:
updating traces for appinsights node test

surgupta-msft committed:
Adding comments

fabiocav commented:
@vrdmr we're getting close to merging this, but I wanted to see if you have any additional feedback you want to provide.

alrod commented on src/WebJobs.Script/Host/HostFunctionMetadataProvider.cs:
```diff
@@ -37,8 +38,10 @@ public HostFunctionMetadataProvider(IOptionsMonitor<ScriptApplicationHostOptions
            => _functionErrors.ToImmutableDictionary(kvp => kvp.Key, kvp => kvp.Value.ToImmutableArray());
 
-        public ImmutableArray<FunctionMetadata> GetFunctionMetadata(IEnumerable<RpcWorkerConfig> workerConfigs, bool forceRefresh)
+        public async Task<ImmutableArray<FunctionMetadata>> GetFunctionMetadataAsync(IEnumerable<RpcWorkerConfig> workerConfigs, bool forceRefresh)
```
Do not need to be `async`

surgupta-msft replied:
This is async because its interface IFunctionMetadataProvider has async method for another implementation - WorkerFunctionMetadataProvider with await calls.

alrod commented on src/WebJobs.Script.Grpc/azure-functions-language-worker-protobuf/src/proto/FunctionRpc.proto:
```diff
@@ -77,9 +77,8 @@ message StreamingMessage {
-    WorkerFunctionMetadataRequest worker_function_metadata_request = 29;
-    WorkerFunctionMetadataResponse worker_function_metadata_response = 30;
-    WorkerFunctionMetadata worker_function_metadata = 31;
+    FunctionsMetadataRequest functions_metadata_request = 29;
```
We also need a PR in https://github.com/Azure/azure-functions-language-worker-protobuf

surgupta-msft replied:
These fields are already added to Protobuf repo - https://github.com/Azure/azure-functions-language-worker-protobuf/blob/dev/src/proto/FunctionRpc.proto#:~:text=functions_metadata_request 

There are 2 pending fields which are part of this PR - https://github.com/Azure/azure-functions-language-worker-protobuf/pull/57

alrod commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -117,14 +122,22 @@ private void InitializeServices()
+            var workerConfigs = _languageWorkerOptions.Value.WorkerConfigs;
+
+            IFunctionMetadataProvider metadataProvider = Utility.CanWorkerIndex(workerConfigs, _environment)
+                ? new WorkerFunctionMetadataProvider(_loggerFactory.CreateLogger<WorkerFunctionMetadataProvider>(), dispatcher)
```
We can add something like `IFunctionMetadataProviderFactory` which will create appropriate `IFunctionMetadataProvider` depending on the config. `IFunctionMetadataProviderFactory` is added to `FunctionMetadataManager` using DI.

To get a dispatcher use `FunctionInvocationDispatcherFactory`

alrod commented on src/WebJobs.Script/Host/ScriptHost.cs:
```diff
@@ -370,9 +376,19 @@ private void LogHostFunctionErrors()
         {
-            IEnumerable<FunctionMetadata> functionMetadata = _functionMetadataManager.GetFunctionMetadata();
+            IEnumerable<FunctionMetadata> functionMetadata;
+            if (workerIndexing)
```
Do not need this if we will have `IFunctionMetadataProvider`

alrod commented on src/WebJobs.Script/Host/IFunctionMetadataManager.cs:
```diff
@@ -3,14 +3,15 @@
         ImmutableDictionary<string, ImmutableArray<string>> Errors { get; }
 
-        ImmutableArray<FunctionMetadata> GetFunctionMetadata(bool forceRefresh = false, bool applyAllowlist = true, bool includeCustomProviders = true);
+        ImmutableArray<FunctionMetadata> GetFunctionMetadata(bool forceRefresh = false, bool applyAllowlist = true, bool includeCustomProviders = true, IFunctionInvocationDispatcher dispatcher = null);
```
Do not need to change the interface if will have `IFunctionMetadataProviderFactory`.

alrod commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -458,64 +458,72 @@ internal async Task SendInvocationRequest(ScriptInvocationContext context)
         }
 
         // gets metadata from worker
-        public Task<List<FunctionMetadata>> WorkerGetFunctionMetadata()
```
Was this used before? I do not see how we call it in the old code.

surgupta-msft replied:
This method is present in Dev branch but was not getting used. In this PR it is getting referenced at multiple places.

alrod commented on src/WebJobs.Script.Grpc/azure-functions-language-worker-protobuf/src/proto/FunctionRpc.proto:
```diff
@@ -265,51 +264,36 @@ message RpcFunctionMetadata {
 }
 
 // Worker sends function metadata back to host
-message WorkerFunctionMetadataResponse {
```
Did any language worker supported this before?

surgupta-msft replied:
This field is not present in Protobuf repo - dev branch. This seems to be added in function-host repo as part of worker-indexing changes and is now getting renamed to FunctionMetadataResponses. This renamed field is there in protobuf dev branch - https://github.com/Azure/azure-functions-language-worker-protobuf/blob/dev/src/proto/FunctionRpc.proto#:~:text=FunctionsMetadataRequest%20functions_metadata_request%20%3D%2029%3B-,FunctionMetadataResponses,-function_metadata_responses%20%3D%2030%3B

alrod commented on src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:
```diff
@@ -266,11 +285,62 @@ public async Task InitializeAsync(IEnumerable<FunctionMetadata> functions, Cance
+        // Gets metadata from worker
+        public async Task<IEnumerable<RawFunctionMetadata>> GetWorkerMetadata()
+        {
+            var channels = (await GetAllWorkerChannelsAsync()).ToArray();
```
Do we need to call `GetInitializedWorkerChannelsAsync` here?

surgupta-msft replied:
Done. Thanks.

alrod commented on test/WebJobs.Script.Tests.Integration/ApplicationInsights/ApplicationInsightsCSharpEndToEndTests.cs:
```diff
@@ -16,7 +16,7 @@ public class TestFixture : ApplicationInsightsTestFixture
             private const string ScriptRoot = @"TestScripts\CSharp";
 
-            public TestFixture() : base(ScriptRoot, "csharp")
+            public TestFixture() : base(ScriptRoot, "dotnet")
```
why did we change this? App Insights is not related to the change.

surgupta-msft replied:
This change is done to fix the AppInsights integration tests failure. The worker runtime env variable was not getting assigned properly causing Node tests to fail. If I set the env variable as shown in the link below, then dotnet tests are failing as it does not understand "csharp" as runtime.

https://github.com/Azure/azure-functions-host/blob/b860e31f0d02a3fbc7f55e8b287b4665f6e07123/test/WebJobs.Script.Tests.Integration/ApplicationInsights/ApplicationInsightsTestFixture.cs#:~:text=Path.GetTempPath()%2C%20%40%22Functions%22)%3B-,Environment,-.SetEnvironmentVariable(EnvironmentSettingNames.FunctionWorkerRuntime

alrod commented on src/WebJobs.Script/Host/ScriptHost.cs:
```diff
@@ -313,15 +319,15 @@ public async Task InitializeAsync(CancellationToken cancellationToken = default)
-                var filteredFunctionMetadata = functionMetadataList.Where(m => m.IsProxy() || !Utility.IsCodelessDotNetLanguageFunction(m));
-                await _functionDispatcher.InitializeAsync(Utility.GetValidFunctions(filteredFunctionMetadata, Functions), cancellationToken);
-
+                if (!workerIndexing)
```
Can we check _functionDispatcher.State instead?

surgupta-msft replied:
workerIndexing variable also takes FeatureFlag into consideration.

alrod commented:
Can you please rebase on latest dev

surgupta-msft committed:
AppInsights tests correction

surgupta-msft committed:
Merge branch 'dev' into t-anjanan/workingSteinChanges-fork

surgupta-msft committed:
ApplicationInsights test fixing

surgupta-msft committed:
Merge branch 'dev' into t-anjanan/workingSteinChanges-fork

surgupta-msft commented:
> Can you please rebase on latest dev

yes, last done on 11/8. Doing this regularly.

surgupta-msft committed:
using GetInitializedWorkerChannelsAsync instead of GetAllWorkerChannelsAsync
